x509cert/encode.go

```diff
@@ -43,25 +45,137 @@ func DecodeRequest(req *proto.X509CertificateSigningRequest) (*x509.Certificate,
+		switch eku {
+		case x509.ExtKeyUsageAny:
+			oid = oidExtKeyUsageAny
+		case x509.ExtKeyUsageServerAuth:
+			oid = oidExtKeyUsageServerAuth
+		case x509.ExtKeyUsageClientAuth:
+			oid = oidExtKeyUsageClientAuth
+		case x509.ExtKeyUsageCodeSigning:
+			oid = oidExtKeyUsageCodeSigning
+		case x509.ExtKeyUsageEmailProtection:
+			oid = oidExtKeyUsageEmailProtection
+		case x509.ExtKeyUsageIPSECEndSystem:
+			oid = oidExtKeyUsageIPSECEndSystem
+		case x509.ExtKeyUsageIPSECTunnel:
+			oid = oidExtKeyUsageIPSECTunnel
+		case x509.ExtKeyUsageIPSECUser:
+			oid = oidExtKeyUsageIPSECUser
+		case x509.ExtKeyUsageTimeStamping:
+			oid = oidExtKeyUsageTimeStamping
+		case x509.ExtKeyUsageOCSPSigning:
+			oid = oidExtKeyUsageOCSPSigning
+		case x509.ExtKeyUsageMicrosoftServerGatedCrypto:
+			oid = oidExtKeyUsageMicrosoftServerGatedCrypto
+		case x509.ExtKeyUsageNetscapeServerGatedCrypto:
+			oid = oidExtKeyUsageNetscapeServerGatedCrypto
+		case x509.ExtKeyUsageMicrosoftCommercialCodeSigning:
+			oid = oidExtKeyUsageMicrosoftCommercialCodeSigning
+		case x509.ExtKeyUsageMicrosoftKernelCodeSigning:
+			oid = oidExtKeyUsageMicrosoftKernelCodeSigning
+		}
+		// Group all EKUs based on whether Critical flag is set.
+		// Each set forms a separate pkix.Extension.
+		if critMap[eku] {
+			extCritTrue = append(extCritTrue, oid)
+		} else {
+			extCritFalse = append(extCritFalse, oid)
+		}
+	}
+	if len(extCritTrue) != 0 {
+		oidM, err := asn1.Marshal(extCritTrue)
+		if err != nil {
+			return nil, err
+		}
+		extension := pkix.Extension{
+			Id:       oidExtensionExtendedKeyUsage,
+			Critical: true,
+			Value:    oidM,
+		}
+		extensions = append(extensions, extension)
+	}
+	if len(extCritFalse) != 0 {
+		oidM, err := asn1.Marshal(extCritFalse)
+		if err != nil {
+			return nil, err
+		}
+		extension := pkix.Extension{
+			Id:       oidExtensionExtendedKeyUsage,
+			Critical: false,
+			Value:    oidM,
+		}
+		extensions = append(extensions, extension)
+	}
+	return extensions, nil
+
+}
+
+// getX509ExtKeyUsage returns []pkix.Extension from []int32
+func getX509ExtKeyUsage(x509ExtKeyUsages []int32) ([]pkix.Extension, error) {
+
 	if len(x509ExtKeyUsages) == 0 {
-		return []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth}, nil
+		return extKeyUsageToExtension([]x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth})
 	}
-	var x509ExtKeyUsagesRet []x509.ExtKeyUsage
-	for _, x509ExtKeyUsage := range x509ExtKeyUsages {
-		// validate if extUsage falls under https://golang.org/src/crypto/x509/x509.go?s=18153:18173#L558
-		if x509ExtKeyUsage < 0 || x509ExtKeyUsage > 11 {
-			return nil, fmt.Errorf("invalid x509 ExtKeyUsage value: %d, valid values are [0,1,...11]", x509ExtKeyUsage)
+	ekuRet := make([]x509.ExtKeyUsage, len(x509ExtKeyUsages))
+	for i, eku := range x509ExtKeyUsages {
+		// extKeyUsage valid values - https://golang.org/src/crypto/x509/x509.go?s=18153:18173#L621
+		if eku < 0 || eku > 13 {
```

Comment:
maybe we should turn `13` into a symbolic constant like `MAX_VALID_EKU`, to change it in one place (or at least in less places)?